PERF: HDFStore __unicode__ method

Kiv opened the pull request:
HDFStore __unicode__ now only returns file path info. New info() method has the previous behavior of __unicode__.

```python
import pandas as pd
store = pd.HDFStore('test.h5', 'w')
for i in range(5000):
    store.put('table_{}'.format(i), pd.DataFrame([i]))

# Before
%time str(store)
CPU times: user 26.1 s, sys: 156 ms, total: 26.2 s
Wall time: 26.2 s

# After
%time str(store)
CPU times: user 0 ns, sys: 0 ns, total: 0 ns
Wall time: 40.1 µs
```

 - [x] closes #16503
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

cmohl2013 committed:
ENH: added margins_name parameter for crosstab (#16489)

* ENH #15972 added margins_name parameter for crosstab

* ENH 15972 minor changes as suggested by reviewers

* ENH 15972 correction in whatsnew

* ENH 15972 style changes in whatsnew

Kiv committed:
PERF: HDFStore has faster __unicode__, new info() method with old behavior.

__unicode__ now only returns file path info, not (expensive) details on all existing keys.

jreback commented on pandas/io/pytables.py:
```diff
@@ -1161,6 +1136,35 @@ def copy(self, file, mode='w', propindexes=True, keys=None, complib=None,
         return new_store
 
+    def info(self):
+        """return detailed information on the store"""
```
add a versionadded tag.

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -75,7 +75,7 @@ Removal of prior version deprecations/changes
 Performance Improvements
 ~~~~~~~~~~~~~~~~~~~~~~~~
-
+- :class:`pandas.HDFStore`'s string representation is now faster and less detailed. For the previous behavior, use ``pandas.HDFStore.info()``. (:issue:`16503`).
```
move to API breaking changes

jreback commented on pandas/io/pytables.py:
```diff
@@ -1161,6 +1136,35 @@ def copy(self, file, mode='w', propindexes=True, keys=None, complib=None,
+    def info(self):
+        """return detailed information on the store"""
+        output = '%s\nFile path: %s\n' % (type(self), pprint_thing(self._path))
+        if self.is_open:
```
add this to api.rst

jreback reviewed (approved):
lgtm, with some minor comment.

jreback commented:
ping when fixed / green.

abarber4gh committed:
TST: ujson tests are not being run (#16499) (#16500)

closes #16499

gfyoung committed:
DOC: Remove preference for pytest paradigm in assert_raises_regex (#16518)

gfyoung commented:
1) Got a small linting error in `pytables.py` (hence the Travis failure)

2) Can you provide some timing info in the PR given that this is a performance change?

Kiv commented:
@jreback I think this is ready to go?

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -4371,11 +4373,11 @@ def test_multiple_open_close(self):
             # single
             store = HDFStore(path)
-            assert 'CLOSED' not in str(store)
+            assert 'CLOSED' not in store.info()
```
so you should also test str(store) a bit (u changed all of these to .info())

Kiv replied:
I added a test that str(store) prints the filename, is that what you mean? There's not much behavior in there anymore to test...

jreback replied:
point to it it's not clear where it is

Kiv replied:
https://github.com/pandas-dev/pandas/pull/16514/commits/6faa5a66c005343c71b1ce623bb3ff89affd72b1#diff-5066b062c892df33c29b37a8c1c75f75R422

neirbowj committed:
TST: Specify HTML file encoding on PY3 (#16526)

gfyoung commented:
@Kiv Some timing info in the PR would be useful for future reference.  Plus a benchmark test would be good here.

TomAugspurger committed:
BUG: Fixed tput output on windows (#16496)

keitakurita committed:
BUG: Incorrect handling of rolling.cov with offset window (#16244)

TomAugspurger committed:
TST: Avoid global state in matplotlib tests (#16539)

Replaces most uses of implicit global state from matplotlib in
test_datetimelike.py. This was potentially causing random failures
where a figure expected to be on a new, blank figure would instead
plot on an existing axes (that's the guess at least).

VincentLa committed:
DOC: Update to docstring of DataFrame(dtype) (#14764) (#16487)

* Adding some more documentation on dataframe with regards to dtype

* Making example for creating dataframe from np matrix easier

ProsperousHeart committed:
DOC: correct docstring examples (#3439) (#16432)

jtratner committed:
Fix unbound local with bad engine (#16511)

Tafkas committed:
return empty MultiIndex for symmetrical difference on equal MultiIndexes (#16486)

JosephWagner committed:
BUG: select_as_multiple doesn't respect start/stop kwargs GH16209 (#16317)

jreback committed:
BUG: Bug in .resample() and .groupby() when aggregating on integers (#16549)

closes #16361

mattip committed:
COMPAT: cython str-to-int can raise a ValueError on non-CPython (#16563)

pepicello committed:
CLN: raise correct error for Panel sort_values (#16532)

TomAugspurger committed:
BUG: Fixed pd.unique on array of tuples (#16543)

pvomelveny committed:
BUG: Allow non-callable attributes in aggregate function. Fixes GH16405 (#16458)

TomAugspurger committed:
Strictly monotonic (#16555)

gfyoung committed:
COMPAT: Consider Python 2.x tarfiles file-like (#16533)

CRP committed:
BUG: Fixed to_html ignoring index_names parameter

closes #16493

Author: Christian Prinoth <christian@prinoth.name>
Author: Tom Augspurger <tom.w.augspurger@gmail.com>
Author: Christian Prinoth <c.prinoth@CPmbpro.local>
Author: Jeff Reback <jeff.reback@twosigma.com>

This patch had conflicts when merged, resolved by
Committer: Jeff Reback <jeff.reback@twosigma.com>

Closes #16495 from CRP/bugfix_16493 and squashes the following commits:

567ae69 [Jeff Reback] doc corrections
8429f9a [Tom Augspurger] Fixed lint error
469a0e6 [Christian Prinoth] BUG: fix for bug 16493
20d512f [Christian Prinoth] BUG: fix for bug 16493
6bef829 [Christian Prinoth] BUG: fix for bug 16493
426565e [Christian Prinoth] BUG: fix for bug 16493
a40820d [Christian Prinoth] BUG: fix for bug 16493

economy committed:
BUG: fixed wrong order of ordered labels in pd.cut()

closes #16459

Author: economy <the.economy@gmail.com>

This patch had conflicts when merged, resolved by
Committer: Jeff Reback <jeff.reback@twosigma.com>

Closes #16466 from economy/fix_cut and squashes the following commits:

29128b3 [economy] comments and whatsnew edits
3898b72 [economy] BUG: fixed wrong order of ordered labels in pd.cut()

jreback committed:
fix linting

Commit:
TST: writing invalid table names to sqlite (#16464)

* Add test for bug #13206.

* Improve test by reading back the values from sql and comparing. Also fixes coding style violation.

neirbowj committed:
TST: Skip test_database_uri_string if pg8000 importable (#16528)

sullivanrs committed:
DOC: Remove incorrect elements of PeriodIndex docstring (#16553)

* DOC: Remove incorrect elements of PeriodIndex docstring

See #9056.

* Removed trailing space

TomAugspurger committed:
TST: Make HDF5 fspath write test robust (#16575)

The test_write_fspath_all test would fail on the HDF5 example
occasionally (about 1/100 in my experience). Apparently you don't get an
identical HDF5 every single time. This refactors that test out to its own where
we write and read both versions, and compare equality that way.

dsm054 committed:
ENH: add .ngroup() method to groupby objects (#14026) (#14026)

OlegShteynbuk committed:
make null lowercase a missing value (#16534)

gfyoung committed:
MAINT: Drop has_index_names input from read_excel (#16522)

jreback commented:
@Kiv lgtm. if you would add an asv would be great (you can use your original example, just not as long, maybe 20 nodes or something would be fine).

rhendric committed:
BUG: reimplement MultiIndex.remove_unused_levels (#16565)

chrisgorgo committed:
Adding 'n/a' to list of strings denoting missing values (#16079)

TomAugspurger committed:
API: Make is_strictly_monotonic_* private (#16576)

jorisvandenbossche committed:
DOC: change doc build to python 3.6 (#16545)

* DOC: change doc build to python 3.6

* Remove pinning of pyqt to 4.x

* Remove pinning of openpyxl

* Add xsel to doc build for clipboard

jreback committed:
DOC: whatsnew 0.20.2 edits (#16587)

funnycrab committed:
DOC: Fix typo in timeseries.rst (#16590)

TomAugspurger committed:
PERF: vectorize _interp_limit (#16592)

* PERF: vectorize _interp_limit

* CLN: remove old implementation

* fixup! CLN: remove old implementation

benjello committed:
DOC: Fix typo in merge doc for validate kwarg (#16595)

makmanalp committed:
BUG: convert numpy strings in index names in HDF #13492 (#16444)

* BUG: Handle numpy strings in index names in HDF5 #13492

* REF: refactor to _ensure_str

bpraggastis committed:
ERRR: Raise error in usecols when column doesn't exist but length matches (#16460)

* gh-14671 Check if usecols with type string contains a subset of names, if not throws an error

* tests added for gh-14671, expected behavior of simultaneous use of usecols and names unclear so these tests are commented out

* Review comments

TomAugspurger committed:
DOC: Whatsnew fixups (#16596)

TomAugspurger committed:
DOC: Update release.rst

jreback committed:
BUG: pickle compat with UTC tz's (#16611)

closes #16608

jhelie committed:
Fix some lgtm alerts (#16613)

jreback committed:
BLD: fix numpy on 3.6 build as 1.13 was released but no deps are built for it (#16633)

dsm054 committed:
BUG: Fix Series.get failure on missing NaN (#8569) (#16619)

dsm054 committed:
TST: NaN in MultiIndex should not become a string (#7031) (#16625)

dsm054 committed:
TST: verify we can add and subtract from indices (#8142) (#16629)

preddy5 committed:
BUG: conversion of Series to Categorical (#16557)

fix #16524

jreback committed:
BLD: fix numpy on 2.7 build as 1.13 was released but no deps are built for it (#16633) (#16650)

tswast committed:
CLN: make license file machine readable (#16649)

Splits extra information about the license and copyright holders to
AUTHORS.md.

jreback committed:
fix pytest-xidst version as 1.17 appears buggy (#16652)

jreback committed:
COMPAT: numpy 1.13 test compat (#16654)

* COMPAT: numpy 1.13 test compat

* CI: fix doc build to 1.12

jreback committed:
Revert "fix pytest-xidst version as 1.17 appears buggy (#16652)" (#16657)

This reverts commit ec6bf6deaf502ac05a7120df13bd9b13cb3083f6.

1.17.1 released that fixes

jreback commented:
can you add an asv?

Kiv committed:
Add ASV benchmark.

Kiv commented:
My git branch got out of date and I don't know how to update it properly. New version with ASV bench is here:

https://github.com/pandas-dev/pandas/pull/16666

jreback commented:
in the future see: http://pandas.pydata.org/pandas-docs/stable/contributing.html#combining-commits

you just needed to do:

```
git rebase -i upstream/master
git push yourremove yourbranch -f
```
